chore: Add target specification (#11477)
vectordotdev/vector: docs/specs/target.md

  * SHOULD deploy across 2 availability zones, MUST be overridable by the user.
* Sizing
  * MUST deploy in a way that takes full advantage of all system resources.
    The Vector service should not be artificially limited with resource
    limiters such as cgroups.
  * SHOULD request 8 vCPUs, MUST be overridable by the user.
  * SHOULD request 2 GiB of memory per vCPU (16 GiB in this case), MUST be
    overridable by the user.
  * SHOULD request 6 GiB of disk space per vCPU (48 GiB in this case), MUST be
    overridable by the user.
  * SHOULD use the following instances if deployed in a listed cloud.
    * AWS SHOULD default to `c6g.2xlarge` instances with 48 GiB of EBS `io2`
      disk space.
    * Azure SHOULD default to `f8` instances with 48 GiB of standard SSD disk
      space.
    * GCP SHOULD default to `c2` instances with 8 vCPUS, 16 GiB of memory, and
      48 GiB of SSD persisted disk space.
* Scaling
  * Load-balancing?...
  * Autoscaling SHOULD be enabled by default driven by an 85% CPU utilization
    target over a rolling 5 minute window.
  * Autoscaling SHOULD have a stabilization period of 5 minutes.

### 5. Unified Architecture

> TODO: Should we support this as a top-level architecture, or have users deploy

Reviewer: As a top-level arch: This would be awkward to accomplish with the current chart design and possibly with upgrade processes down the line. I'd lean towards two deployments.

both the agent and aggregator separately and integrate them by default?

## 6. Hardening

* Setup
  * An unprivileged Vector service account SHOULD be created upon installation
    for running the Vector process.
* Data hardening
  * Swap SHOULD be disabled to prevent in-flight data from leaking to disk.
    Swap would also make Vector prohibitively slow.
  * Vector's data directory SHOULD be read and write restricted to Vector's
    dedicated service account.
  * Core dumps SHOULD be prevented for the Vector process to prevent in flight
    data from leaking to disk.
* Process hardening
  * Vector's artifacts
    * All communication during the setup process, such as downloading Vector
      artifacts, MUST use encrypted channels.
    * Downloaded Vector artifacts MUST be verified against the provided
      checksum.
    * The latest Vector version SHOULD be downloaded unless otherwise specified
      by the user.
  * Vector's configuration
    * Vector's configuration directory SHOULD be read restricted to Vector's
      service account.
  * Vector's runtime
    * Vector MUST be run under an unprivileged, deciated service account by

>       default.

Author: Yeah, we might consider splitting the spec on the deployment roles/architectures. I'm doing that with the docs now and it's helpful.

>       default.

Author: Then let's move it under the aggregator section and make it exclusive to that architecture.

# Target Specification

> This document specifies requirements for deployment targets for the

Reviewer: Broadly we can split our targets between `Virtual/Physical Machines` and `Orchestrators`.

- `Virtual/Physical Machines`
  - Target is a single instance.
  - This implies limited, to no, _High Availability_ and _Scaling_ spec adherence.
  - However, there should be no issues with general _Architecture_, _Sizing_, and _Hardening_.
- `Orchestrators`
  - Target is a group of nodes, managed by a scheduler or orchestrator.
  - A management layer (Kubernetes, Nomad, Swarm, Mesos, DC/OS) over a group of nodes allows us to adhere to _High Availability_ and _Scaling_ spec more easily.
  - These targets should expose primitives to ensure the _Architecture_, _Sizing_, and _Hardening_ spec is met.

There is some space between these two classes for things like Chef, Puppet, Ansible, or even Terraform that could help bridge this gap but we do not currently provide any resources for these types of deployments.

node for local data collection. Along with general [hardening](#6-hardening)
requirements, the following requirements define support for this architecture:

- Architecture
  - MUST deploy as a daemon on existing nodes, one Vector process per node.
  - MUST deploy with Vector's [default agent configuration][default_agent_configuration]
    which largely covers the agent architecture
    [design recommendations][agent_architecture].
- Sizing
  - MUST deploy as a good infrastructure citizen, giving resource priority to
    other services on the same node.
  - SHOULD be limited to 2 vCPUs, MUST be overridable by the user.
  - SHOULD be limited to 2 GiB of memory per vCPU (4 GiB in this case), MUST be
    overridable by the user.
  - SHOULD be limited to 1 GiB of disk space, MUST be overridable by the user.

### 4. Aggregator Architecture

The [aggregator architecture][aggregator_architecture] deploys Vector onto
dedicated nodes for data aggregation. Along with general [hardening](#6-hardening)
requirements, the following requirements define support for this architecture:

- Architecture
  - SHOULD deploy as a service on dedicated nodes, Vector is the only service
    on the node.

>   - MUST deploy with a persistent disk that is available between deployments.

Reviewer 2: I think `SHOULD` is still good here. If users are using aggregators with something like Kafka or SQS than a disk is unnecessary. That is: I think we should still support Stateless Aggregators in k8s.

- Process hardening
  - Vector's artifacts
    - All communication during the setup process, such as downloading Vector
      artifacts, MUST use encrypted channels.
    - Downloaded Vector artifacts MUST be verified against the provided
      checksum.
    - The latest Vector version SHOULD be downloaded unless otherwise specified
      by the user.
  - Vector's configuration
    - Vector's configuration directory SHOULD be read restricted to Vector's
      service account.
  - Vector's runtime
    - Vector SHOULD be run under an unprivileged, dedicated service account.
    - Vector's service account SHOULD NOT have the ability to overwrite Vector's
      binary or configuration files. The only directory the Vector service
      account should write to is Vector’s data directory.
- Network hardening
  - Configured sources and sinks SHOULD use encrypted channels by default.

[agent_architecture]: https://vector.dev/docs/setup/going-to-prod/arch/agent/
[aggregator_architecture]: https://vector.dev/docs/setup/going-to-prod/arch/aggregator/
[default_agent_configuration]: https://github.com/vectordotdev/vector/blob/target-specification/config/agent/vector.yaml

> [default_aggregator_configuration]: https://vector.dev/docs/reference/configuration/default/aggregator.yaml

Reviewer: Not sure where we want the default configs to live, or at least where to link to them.

Author: Somewhere in the the `config` directory in the Vector repo. So `config/agent.yaml` and `config/aggregator.yaml`, or `config/defaults/*`/

[hardening]: https://vector.dev/docs/setup/going-to-prod/hardening/
[high_availability]: https://vector.dev/docs/setup/going-to-prod/high-availability/
[reference_architectures]: https://vector.dev/docs/setup/going-to-prod/arch/
[RFC 2119]: https://datatracker.ietf.org/doc/html/rfc2119

> [terminology_document]: https://vector.dev/docs/reference/glossary/

Reviewer: This is the most terminology document I could find that we publish, but it's lacking a number of definitions - specifically the mentioned `"agent", "aggregator", "node", and "service"`.

Author: That's fine for now. Yes, it will be in the reference on the new site and updated.

allow for a scheduler to deploy Vector across a number nodes. These categories
have their own requirements for each
[Deployment Architecture](#3-deployment-architectures).

Examples of Virtual/Physical Machine targets include, but are not limited to:

- Debian
- Docker
- RHEL
- Windows

Examples of Orchestration Platform targets include, but are not limited to:

- Kubernetes

## 3. Deployment Architectures

When supporting a target, Vector must support them through the the paradigm of
architectures:

- Targets MUST support the [agent architecture][agent_architecture] by
  providing a single command that deploys Vector and achieves the
  [agent architecture requirements](#agent-architecture).
- Targets SHOULD support the [aggregator architecture][aggregator_architecture] by
  providing a single command that deploys Vector and achieves the
  [aggregator architecture requirements](#aggregator-architecture).
- Targets MAY support the [unified architecture][unified_architecture] by
  providing a single command that deploys Vector and achieves the
  [unified architecture requirements](#unified-architecture).

### 4. Agent Architecture

> The [agent architecture][agent_architecture] deploys Vector on each individual
> node for local data collection. Along with general [hardening](#7-hardening)
> requirements, the following requirements define support for this architecture:

Author: ```suggestion
The [agent architecture][agent_architecture] deploys Vector on each individual
node for distributed data collection and processing. Along with general
[hardening](#7-hardening) requirements, the following requirements define support
for this architecture:
```

- Architecture

>   - MUST deploy as a daemon on existing nodes, one Vector process per node.

Author: ```suggestion
  - MUST deploy as a daemon on existing nodes, not a sidecar -- one Vector process per physical node.
```

>   - MUST deploy with Vector's [default agent configuration][default_agent_configuration]
>     which largely covers the agent architecture
>     [design recommendations][agent_architecture].

Author: ```suggestion
  - MUST deploy with Vector's [default agent configuration][default_agent_configuration].
```

- Sizing
  - MUST deploy as a good infrastructure citizen, giving resource priority to
    other services on the same node.

>   - SHOULD be limited to 2 vCPUs, MUST be overridable by the user.

Author: ```suggestion
  - SHOULD be limited to 1 vCPUs by default, MUST be overridable by the user.
```

>   - SHOULD be limited to 2 GiB of memory per vCPU (4 GiB in this case), MUST be
>     overridable by the user.

Author: ```suggestion
  - SHOULD be limited to 2 GiB of memory per vCPU by default, MUST be
    overridable by the user.
```

  - SHOULD be limited to 1 GiB of disk space, MUST be overridable by the user.

### 5. Aggregator Architecture

The [aggregator architecture][aggregator_architecture] deploys Vector onto
dedicated nodes for data aggregation. Along with general [hardening](#7-hardening)
requirements, the following requirements define support for this architecture:

- Architecture
  - SHOULD deploy as a service on dedicated nodes, Vector is the only service
    on the node.

>   - MUST deploy with a persistent disk that is available between deployments.

Author: ```suggestion
  - SHOULD deploy with a persistent disk that is available between deployments by default,
     MUST be overridable by the user if they do not want a persistent disk.
```

  - MUST deploy with Vector's [default aggregator configuration][default_aggregator_configuration]
    which largely covers the aggregator architecture
    [design recommendations][aggregator_architecture].
    in order to achieve durability with disk buffers and source checkpoints.
  - SHOULD deploy within one Cluster or VPC at a time.
  - Configured Vector ports, including non-default user configured ports,

>     SHOULD be automatically accessible within the Cluster or VPC.

Author: That's pretty cool. I'm a fan of it if the complexity doesn't outweigh the benefit.

>     SHOULD be automatically accessible within the Cluster or VPC.

Reviewer: 👍 it _is_ pretty cool if I do say so myself 😆 Being able to do it where possible, but leaving an escape hatch for users seems reasonable to me. There's definitely not a lot of places we'd be able to do it.

  - Configured Vector sources, including non-default user configured sources,
    SHOULD be automatically discoverable via target service discovery
    mechanisms.
- Sizing
  - MUST deploy in a way that takes full advantage of all system resources.
  - The Vector service SHOULD NOT be artificially limited with resource

>     limiters such as cgroups.

Reviewer: We can set a `request` that ensures we're on a node with _at least_ that much resource available, and not set a `limit` (which handles throttling/oomkilling Vector)

>     limiters such as cgroups.

Reviewer 2: Indeed, I figured systemd would make use of cgroups too.

>     limiters such as cgroups.

Reviewer: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/ documents detailing how this works in Kube land

>   - SHOULD request 8 vCPUs, MUST be overridable by the user.

Author: ```suggestion
  - SHOULD require 8 vCPUs by default, MUST be overridable by the user.
```

  - SHOULD request 2 GiB of memory per vCPU (16 GiB in this case), MUST be
    overridable by the user.

>   - SHOULD request 6 GiB of disk space per vCPU (48 GiB in this case), MUST be
>     overridable by the user.

Author: ```suggestion
  - SHOULD request 36 GiB of disk space per vCPU by default (288 GiB in this case),
     MUST be overridable by the user.
```

The following are additional requirements for Orchestration Platform installation
targets:

- High Availability
  - SHOULD deploy across multiple nodes, MUST be overridable by the user.
  - SHOULD deploy across multiple availability zones, MUST be overridable by the user.
- Scaling

>   - MUST deploy along with a load balancer to enable horizontal autoscaling.

Author: ```suggestion
  - SHOULD provide facilities for provisioning a load balancer to enable horizontal scaling
     out of the box. MUST be overridable by the user.
     - Cloud-managed load balancers (i.e., AWS NLB) SHOULD be supported in addition to
        self-managed load balancers (i.e., HAProxy).
     - Cloud-managed load balancers (i.e., AWS NLB) SHOULD be prioritized by default over
        self-managed load balancers (i.e., HAProxy).
     - Network load balancers (layer-4) SHOULD be prioritized over HTTP load balancers (layer-7)
```

# Target Specification

This document specifies requirements for installation targets for the
integration of Vector.

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”,
“SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be
interpreted as described in [RFC 2119].

Other words, such as "agent", "aggregator", "node", and "service" are to be
interpreted as described in the [terminology document][terminology_document].

- [1. Introduction](#1-introduction)
- [2. Installation Targets](#2-installation-targets)
- [3. Deployment Architectures](#3-deployment-architectures)
  - [4. Agent Architecture](#4-agent-architecture)
  - [5. Aggregator Architecture](#5-aggregator-architecture)
  - [6. Unified Architecture](#6-unified-architecture)
- [7. Hardening](#7-hardening)

## 1. Introduction

In its simplest form, installing Vector consists of downloading the binary and
making it executable, but leaves much to be desired for users looking to
integrate Vector in real-world production environments. To adhere with Vector's

> ["reduce decisions" design principle], Vector must also be opinionated about how

Reviewer 3: > ["reduce decisions" design principle]

Was this intended to be a link somewhere?

Reviewer: Oops that got dropped between commits somehow - was intended to link to https://github.com/vectordotdev/vector/blob/master/docs/USER_EXPERIENCE_DESIGN.md#be-opinionated--reduce-decisions

Reviewer: Fixed in f3f68eec90bbb4568db9023c733ee8c47e368ec7

it's deployed, providing easy facilities for adopting Vector's
[reference architectures][reference_architectures],
[achieving high availability][high_availability], and [hardening][hardening]
Vector.

## 2. Installation Targets

Vector supports a number of installation targets that can be categorized into:

- Virtual/Physical Machine
- Orchestration Platform

The primary differentiator between the two being that Virtual/Physical Machines
provide a single node as the deployment target, whereas Orchestration Platforms

> allow for a scheduler to deploy Vector across a number nodes. These categories

Reviewer 4: ```suggestion
allow for a scheduler to deploy Vector across a number of nodes. These categories
```

requirements, the following requirements define support for this architecture:

- Architecture
  - MUST deploy as a service with reserved/dedicated resources.
  - SHOULD deploy with a persistent disk that is available between deployments by default,
     MUST be overridable by the user if they do not want a persistent disk.
  - MUST deploy with Vector's [default aggregator configuration][default_aggregator_configuration].
  - Configured Vector ports, including non-default user configured ports,
    SHOULD be automatically accessible within the Cluster or VPC.
  - Configured Vector sources, including non-default user configured sources,
    SHOULD be automatically discoverable via target service discovery
    mechanisms.
- Sizing
  - MUST have dedicated/reserved resources that cannot be stolen by other services, preventing
     the "noisy neighbor" problem to the degree possible.
  - The Vector service SHOULD NOT be artificially limited with resource
    limiters such as cgroups.
  - SHOULD require 8 vCPUs by default, MUST be overridable by the user.
  - SHOULD require 2 GiB of memory per vCPU (16 GiB in this case) by default,
     MUST be overridable by the user.

>   - SHOULD request 36 GiB of disk space per vCPU by default (288 GiB in this case),

Reviewer 3: Real glad for clarification on this one, I'll update my POV doc to reflect this.

     MUST be overridable by the user.

The following are additional requirements for Orchestration Platform installation
targets:

- High Availability
  - SHOULD deploy across multiple nodes by default, MUST be overridable by the user.
  - SHOULD deploy across multiple availability zones by default, MUST be overridable by the user.
- Scaling
  - SHOULD provide facilities for provisioning a load balancer to enable horizontal scaling
     out of the box. MUST be overridable by the user.
     - Cloud-managed load balancers (i.e., AWS NLB) SHOULD be supported in addition to
        self-managed load balancers (i.e., HAProxy).
     - Cloud-managed load balancers SHOULD be prioritized by default over self-managed
        load balancers.
     - Network load balancers (layer-4) SHOULD be prioritized over HTTP load balancers (layer-7)
  - Autoscaling SHOULD be enabled by default, driven by an average of 85%
    CPU utilization and a stabilization period of 5 minutes.

### 6. Unified Architecture

> The [unified architecture][unified_architecture] deploys Vector on each
> individual node for local data collection, as well as onto dedicated nodes for
> data aggregation. The requirements for both the [agent](#4-agent-architecture)
> and the [aggregator](#5-aggregator-architecture) apply to this architecture.
> This architecture SHOULD NOT be installed on Virtual/Physical Machine targets
> as there is little added benefit.

Author: ```suggestion
The [unified architecture][unified_architecture] deploys Vector on each
individual node as an agent and as a separate service as an aggregator.
The requirements for both the [agent](#4-agent-architecture) and the
[aggregator](#5-aggregator-architecture) apply to this architecture.
This architecture SHOULD NOT be installed on Virtual/Physical Machine
targets as there is little added benefit.
```

node for local data collection. Along with general [hardening](#7-hardening)
requirements, the following requirements define support for this architecture:

- Architecture
  - MUST deploy as a daemon on existing nodes, one Vector process per node.
  - MUST deploy with Vector's [default agent configuration][default_agent_configuration]
    which largely covers the agent architecture
    [design recommendations][agent_architecture].
- Sizing
  - MUST deploy as a good infrastructure citizen, giving resource priority to
    other services on the same node.
  - SHOULD be limited to 2 vCPUs, MUST be overridable by the user.
  - SHOULD be limited to 2 GiB of memory per vCPU (4 GiB in this case), MUST be
    overridable by the user.
  - SHOULD be limited to 1 GiB of disk space, MUST be overridable by the user.

### 5. Aggregator Architecture

The [aggregator architecture][aggregator_architecture] deploys Vector onto
dedicated nodes for data aggregation. Along with general [hardening](#7-hardening)
requirements, the following requirements define support for this architecture:

- Architecture
  - SHOULD deploy as a service on dedicated nodes, Vector is the only service

>     on the node.

Reviewer 2: I guess this is a SHOULD for a VM target, but not for k8s?

Reviewer: I think SHOULD is still reasonable - it ought to be done this way unless you understand the implications of not doing it, and have a good reason to ignore the directive. In kubernetes examples it could be binpacking/misc infra tools need to run there/etc.

  - MUST deploy with Vector's [default aggregator configuration][default_aggregator_configuration]
    which largely covers the aggregator architecture
    [design recommendations][aggregator_architecture].
    in order to achieve durability with disk buffers and source checkpoints.
  - SHOULD deploy within one Cluster or VPC at a time.
  - Configured Vector ports, including non-default user configured ports,
    SHOULD be automatically accessible within the Cluster or VPC.
  - Configured Vector sources, including non-default user configured sources,
    SHOULD be automatically discoverable via target service discovery
    mechanisms.
- Sizing
  - MUST deploy in a way that takes full advantage of all system resources.
  - The Vector service SHOULD NOT be artificially limited with resource
    limiters such as cgroups.
  - SHOULD request 8 vCPUs, MUST be overridable by the user.
  - SHOULD request 2 GiB of memory per vCPU (16 GiB in this case), MUST be
    overridable by the user.
  - SHOULD request 6 GiB of disk space per vCPU (48 GiB in this case), MUST be
    overridable by the user.
  - SHOULD use the following instances if deployed in a listed cloud.

>     - AWS SHOULD default to `c6g.2xlarge` instances with 48 GiB of EBS `io2`
>       disk space.

Reviewer 2: ```suggestion
    - AWS SHOULD default to `c6i.2xlarge` instances with 48 GiB of EBS `gp3`
      disk space.
```

I believe this is what we discussed?

Reviewer: This was awaiting the updated going-to-prod docs but I _believe_ those are right.

Reviewer 2: Gotcha, yeah, we can wait and see. I know we want to use `c6i` for now since `c6g` seemed to have worse performance and we don't have soak tests for it yet. I thought we discussed using `gp` disks over `io`, but I'm less sure there.

Reviewer: The more I think about this as a spec - I don't think we should be specifically listing machine types, probably just CPU/RAM/DISK(speed/size)

Reviewer 2: Yeah, I think that'd be fine too.

>     - Azure SHOULD default to `f8` instances with 48 GiB of standard SSD disk
>       space.

Reviewer 2: ```suggestion
    - Azure SHOULD default to `f8` instances with 48 GiB of Standard SSD disk
      space.
```

It wasn't clear to me that "Standard SSD" is an Azure term. We could link to https://docs.microsoft.com/en-us/azure/virtual-machines/disks-types#standard-ssds here too.

Reviewer: See: https://github.com/vectordotdev/vector/pull/11477#discussion_r847344953

Reviewer: Removed in 67ae0d94bffcf0797b17497401e4d7ea2ccc6885

    - GCP SHOULD default to `c2` instances with 8 vCPUS, 16 GiB of memory, and

>       48 GiB of SSD persisted disk space.

Reviewer 2: I assume we mean `pd-ssd` for the disk type here? Could we add that to be extra clear? We could link to https://cloud.google.com/compute/docs/disks#disk-types too.

Reviewer: https://github.com/vectordotdev/vector/pull/11477#discussion_r847344953

### 6. Unified Architecture

The [unified architecture][unified_architecture] deploys Vector on each
individual node for local data collection, as well as onto dedicated nodes for
data aggregation. The requirements for both the [agent](#4-agent-architecture)
and the [aggregator](#5-aggregator-architecture) apply to this architecture.
This architecture SHOULD NOT be installed on Virtual/Physical Machine targets
as there is little added benefit.

## 7. Hardening

- Setup
  - An unprivileged Vector service account SHOULD be created upon installation
    for running the Vector process.
- Data hardening
  - Swap SHOULD be disabled to prevent in-flight data from leaking to disk.
    Swap would also make Vector prohibitively slow.
  - Vector's data directory SHOULD be read and write restricted to Vector's
    dedicated service account.
  - Core dumps SHOULD be prevented for the Vector process to prevent in flight
    data from leaking to disk.
- Process hardening
  - Vector's artifacts
    - All communication during the setup process, such as downloading Vector
      artifacts, MUST use encrypted channels.
    - Downloaded Vector artifacts MUST be verified against the provided

>       checksum.

Reviewer 2: Just noting we don't have these yet.

Reviewer: 👍 should shake out into the get targets to match spec/moving out artifact pipelines onto Datadog
